### ISSUE TITLE
support ip arrays for a replicaset 

### ISSUE THREAD

**Issue by magicsong** [projectcalico/cni-plugin#771]
currently, calico supports fixed ip for only one pod as the docs said: 
`You can only specify one IPv4/IPv6 or one IPv4 and one IPv6 address with these annotations.`
However, in k8s, we seldom use single pod but use replicaset such as deployment or statefulset. 
How do i set fixed ips for such workloads?

**Comment by rafaelvanoni** [comment]
@magicsong Can you try creating an IP pool for each of your deployments and specify that in an annotation? That might address your use case.

**Comment by magicsong** [comment]
yes, it can work. however we have used ippool for a namespace and we cannot use a subset as a pool.

**Comment by tmjd** [comment]
Can you link to the docs you are referring to? I thought when you specify an IP for a pod then the pool must be disabled.
Part of the reasoning here is that you are taking on the role of IP management if you are assigning specific IPs so Calico should not be managing the pool.

**Comment by magicsong** [comment]
@tmjd the doc is here <https://github.com/projectcalico/calico/blob/master/v3.8/reference/cni-plugin/configuration.md#using-kubernetes-annotations>. a bit different

**Comment by tmjd** [comment]
Thanks, I wasn't aware of the annotation `cni.projectcalico.org/ipAddrs`, which sounds like it works with enabled IP Pools.

**Comment by halfcrazy** [comment]
There are two limitations to this annotation. 
1) When using fixed ips, you have to control ipam by yourself as if your deployment/statefulset stopped, those ips are freed and other workloads may take those ips, next time you start up your deployment/statefulset container may not obtain those ips as well. Calico will not preserve those ip for you.
2) If you deploy calico to an on-premise k8s cluster, without node-to-node full mesh or rr. And want to provide pods externally accessibility. Then in the node which has the ipam block will have a blackhole route for ipam block and if any ip within this block is used as fixed ip and appears on other nodes will not be accessible from the node hold this block.

I have a POC to support ip array for fixed ip.
https://github.com/halfcrazy/cni-plugin/commit/c9472f7a0d69c84f98f48983b9207003d4b44d27

**Comment by caseydavenport** [comment]
Also raised here: https://github.com/projectcalico/calico/issues/4893

Fundamentally Kubernetes supports at most one IPv4 and one IPv6 address, which is where this limitation is coming from.

We might be able to circumvent that, but it would come with severe limitations (Kubernetes serivces would only know about one address of each family, for example). So, we need to be careful if we decide to tackle this.

**Comment by logan2211** [comment]
A few days ago I posted patches which accomplish this in a different issue (https://github.com/projectcalico/cni-plugin/issues/429#issuecomment-935630485).. sorry, this issue is more pertinent, but I didn't see it before.

Note that the libcalico-go patch necessitates a patched rebuild of all calico containers. The CNI plugin patch alone will allow calico to support multiple ipAddrs, but the route table will not get set up correctly unless the other calico services contain the libcalico-go patch.
